File: study/lua/src/app/scene/view/pet/PetListCell.lua
--
-- Author: hedili
-- Date: 2018-01-24 13:50:59
--
local ListViewCellBase = require("app.ui.ListViewCellBase")
local PetListCell = class("PetListCell", ListViewCellBase)
local TypeConvertHelper = require("app.utils.TypeConvertHelper")
local UserDataHelper = require("app.utils.UserDataHelper")

local TOPIMAGERES = {
	"img_iconsign_shangzhen", --上阵
	"img_iconsign_shangzhen",--已护佑
}

function PetListCell:ctor()
	local resource = {
		file = Path.getCSB("PetListCell", "pet"),
		binding = {
			_buttonStrengthen1 = {
				events = {{event = "touch", method = "_onButtonStrengthenClicked1"}}
			},
			_buttonStrengthen2 = {
				events = {{event = "touch", method = "_onButtonStrengthenClicked2"}}
			},
		}
	}
	PetListCell.super.ctor(self, resource)
end

function PetListCell:onCreate()
	local size = self._resourceNode:getContentSize()
	self:setContentSize(size.width, size.height)
	
	self._buttonStrengthen1:setString(Lang.get("hero_btn_strengthen"))
	self._buttonStrengthen2:setString(Lang.get("hero_btn_strengthen"))
end

function PetListCell:update(heroId1, heroId2)
	local function updateCell(index, heroId)
		if heroId then
			if type(heroId) ~= "number" then
				return
			end
			self["_item"..index]:setVisible(true)
			local data = G_UserData:getPet():getUnitDataWithId(heroId)
			self["_item"..index]:updateUI(TypeConvertHelper.TYPE_PET, data:getBase_id())
			self["_item"..index]:setCallBack(handler(self, self["_onClickIcon"..index]))
			self:_showTopImage(index, data)

			local icon = self["_item"..index]:getCommonIcon()
			local params = icon:getItemParams()
			local starLevel = data:getStar()
			local name = params.name

			self["_nodeStar"..index]:setCount(starLevel)
			self["_item"..index]:setName(name)
			self["_item"..index]:setTouchEnabled(true)
			self["_nodeLevel"..index]:updateUI(Lang.get("hero_list_cell_level_des"), Lang.get("hero_txt_level", {level = data:getLevel()}))
	
			self["_buttonStrengthen"..index]:setVisible(data:isCanTrain())
		else
			self["_item"..index]:setVisible(false)
		end
	end

	updateCell(1, heroId1)
	updateCell(2, heroId2)
end

function PetListCell:_showTopImage(index, data)
	local imageTop = self["_imageTop"..index]
	local isInBattle = data:isInBattle()
	local isInBless = data:isPetBless()

	if isInBattle then
		imageTop:loadTexture(Path.getTextSignet(TOPIMAGERES[1]))
		imageTop:setVisible(true)
	elseif isInBless then
		imageTop:loadTexture(Path.getTextSignet(TOPIMAGERES[2]))
		imageTop:setVisible(true)
	else
		imageTop:setVisible(false)
	end
end

function PetListCell:_onButtonStrengthenClicked1() 
	self:dispatchCustomCallback(1)
end

function PetListCell:_onButtonStrengthenClicked2()
	self:dispatchCustomCallback(2)
end

function PetListCell:_onClickIcon1(sender, itemParams)
	--if itemParams.cfg.type == 3 then
	--	local PopupItemGuider = require("app.ui.PopupItemGuider").new(Lang.get("way_type_get"))
	--	PopupItemGuider:updateUI(TypeConvertHelper.TYPE_PET, itemParams.cfg.id)
	--	PopupItemGuider:openWithAction()
	--else
		self:dispatchCustomCallback(1)
	--end
end

function PetListCell:_onClickIcon2(sender, itemParams)
	--local PopupItemGuider = require("app.ui.PopupItemGuider").new(Lang.get("way_type_get"))
	--PopupItemGuider:updateUI(TypeConvertHelper.TYPE_PET, itemParams.cfg.id)
	--PopupItemGuider:openWithAction()

	self:dispatchCustomCallback(2)
end

return PetListCell

File: study/ts/script/scene/view/pet/PetListCell.ts
const { ccclass, property } = cc._decorator;

import CommonListCellBase from '../../../ui/component/CommonListCellBase'
import CommonButtonLevel1Highlight from '../../../ui/component/CommonButtonLevel1Highlight';
import CommonDesValue from '../../../ui/component/CommonDesValue';
import CommonHeroStarBig from '../../../ui/component/CommonHeroStarBig';
import CommonListItem from '../../../ui/component/CommonListItem';
import { Lang } from '../../../lang/Lang';
import { handler } from '../../../utils/handler';
import { G_UserData } from '../../../init';
import { TypeConvertHelper } from '../../../utils/TypeConvertHelper';
import { Path } from '../../../utils/Path';
import UIHelper from '../../../utils/UIHelper';

var TOPIMAGERES = [
    'img_iconsign_shangzhen',
    'img_iconsign_shangzhen'
];

@ccclass
export default class PetListCell extends CommonListItem {

    @property({
        type: cc.Node,
        visible: true
    })
    _resourceNode: cc.Node = null;

    @property({
        type: CommonListCellBase,
        visible: true
    })
    _item1: CommonListCellBase = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageTop1: cc.Sprite = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageTop2: cc.Sprite = null;

    @property({
        type: CommonListCellBase,
        visible: true
    })
    _item2: CommonListCellBase = null;

    @property({
        type: CommonButtonLevel1Highlight,
        visible: true
    })
    _buttonStrengthen1: CommonButtonLevel1Highlight = null;

    @property({
        type: CommonButtonLevel1Highlight,
        visible: true
    })
    _buttonStrengthen2: CommonButtonLevel1Highlight = null;

    @property({
        type: CommonDesValue,
        visible: true
    })
    _nodeLevel1: CommonDesValue = null;

    @property({
        type: CommonHeroStarBig,
        visible: true
    })
    _nodeStar1: CommonHeroStarBig = null;

    @property({
        type: CommonDesValue,
        visible: true
    })
    _nodeLevel2: CommonDesValue = null;

    @property({
        type: CommonHeroStarBig,
        visible: true
    })
    _nodeStar2: CommonHeroStarBig = null;


    onLoad() {
        this._buttonStrengthen1.setString(Lang.get('hero_btn_strengthen'));
        this._buttonStrengthen2.setString(Lang.get('hero_btn_strengthen'));

        this._buttonStrengthen1.addClickEventListenerEx(handler(this, this._onButtonStrengthenClicked1));
        this._buttonStrengthen2.addClickEventListenerEx(handler(this, this._onButtonStrengthenClicked2));
    }

    updateUI(index, itemLine) {
        this.updateCell(1, itemLine[0]);
        this.updateCell(2, itemLine[1]);
    }

    updateCell(index, heroId) {
        if (heroId) {
            if (typeof (heroId) != 'number') {
                return;
            }
            this['_item' + index].node.active = (true);
            var data = G_UserData.getPet().getUnitDataWithId(heroId);
            this['_item' + index].updateUI(TypeConvertHelper.TYPE_PET, data.getBase_id());
            this['_item' + index].setCallBack(handler(this, this['_onClickIcon' + index]));
            this._showTopImage(index, data);
            var icon = this['_item' + index].getCommonIcon();
            var params = icon.getItemParams();
            var starLevel = data.getStar();
            var name = params.name;
            this['_nodeStar' + index].setCount(starLevel);
            this['_item' + index].setName(name);
            this['_item' + index].setTouchEnabled(true);
            this['_nodeLevel' + index].updateUI(Lang.get('hero_list_cell_level_des'), Lang.get('hero_txt_level', { level: data.getLevel() }));
            this['_buttonStrengthen' + index].setVisible(data.isCanTrain());
        } else {
            this['_item' + index].node.active = (false);
        }
    }
    _showTopImage(index, data) {
        var imageTop = this['_imageTop' + index];
        var isInBattle = data.isInBattle();
        var isInBless = data.isPetBless();
        if (isInBattle) {
            UIHelper.loadTexture(imageTop, Path.getTextSignet(TOPIMAGERES[0]));
            imageTop.node.active = (true);
        } else if (isInBless) {
            UIHelper.loadTexture(imageTop, Path.getTextSignet(TOPIMAGERES[1]));
            imageTop.node.active = (true);
        } else {
            imageTop.node.active = (false);
        }
    }

    _onButtonStrengthenClicked1() {
        this.dispatchCustomCallback(0);
    }
    _onButtonStrengthenClicked2() {
        this.dispatchCustomCallback(1);
    }
    _onClickIcon1(sender, itemParams) {
        this.dispatchCustomCallback(0);
    }
    _onClickIcon2(sender, itemParams) {
        this.dispatchCustomCallback(1);
    }
}